File: frontend/src/types/skills.ts
export interface Skill {
  id: string;
  name: string;
  category: string;
  level: string;
}

export interface SkillDemand {
  name: string;
  demand: number;
  is_owned: boolean;
}

export interface SuggestedProject {
  title: string;
  description: string;
  tech_stack: string[];
  skills_covered: string[];
  key_steps?: string[];
  deliverables?: string[];
}

export interface SuggestedProjectsCache {
  suggested_projects: SuggestedProject[];
  skills: string[];
  cached: boolean;
}

export interface SkillGap {
  skill: string;
  gap_reason: string;
  priority: "high" | "medium" | "low";
  kind: "missing" | "underleveled";
  current_level?: string | null;
  expected_level?: string | null;
  demand: number;
}

export interface SkillAnalysis {
  core_strengths: string[];
  skill_gaps: SkillGap[];
  target_role_ready: boolean;
  target_role_title?: string | null;
  sample_status?: string | null;
  sample_job_count: number;
  error?: string | null;
}

export interface LearningPhase {
  title: string;
  duration: string;
  goal: string;
  concepts: string[];
  steps: string[];
  project_title?: string | null;
}

export interface LearningRoadmap {
  overview: string;
  overall_duration: string;
  phases: LearningPhase[];
}

export interface LearningRoadmapCache {
  roadmap: LearningRoadmap | null;
  cached: boolean;
}

File: backend/app/schemas/skills.py
from __future__ import annotations

from pydantic import BaseModel


class SkillCreate(BaseModel):
    name: str
    category: str
    level: str


class SkillRead(SkillCreate):
    id: str


class SkillDemand(BaseModel):
    name: str
    demand: int
    # True when the skill is already in the requesting person's own skill
    # set - lets the Market Demand panel mark it "owned" instead of
    # treating every listed skill as something to learn.
    is_owned: bool = False


class SkillGap(BaseModel):
    skill: str
    gap_reason: str
    priority: str
    kind: str = "missing"  # "missing" | "underleveled"
    current_level: str | None = None
    expected_level: str | None = None
    demand: int = 0


class LearningPhase(BaseModel):
    """One ordered step of a :class:`LearningRoadmap`."""

    title: str
    duration: str = ""
    goal: str = ""
    # Atomic study topics (e.g. "chunking strategies"), never vague
    # "read about X" filler - see LEARNING_ROADMAP_PROMPT.
    concepts: list[str] = []
    steps: list[str] = []
    # Title of a SuggestedProject the candidate chose to include, scheduled
    # onto this phase - None when no included project fits.
    project_title: str | None = None


class LearningRoadmap(BaseModel):
    overview: str = ""
    overall_duration: str = ""
    phases: list[LearningPhase] = []


class SuggestedProject(BaseModel):
    """A comprehensive multi-skill project idea from skill-gap analysis."""

    title: str
    description: str = ""
    tech_stack: list[str] = []
    skills_covered: list[str] = []
    key_steps: list[str] = []
    deliverables: list[str] = []


class LearningRoadmapRequest(BaseModel):
    """Generate a roadmap from the unified "skills to learn" chip set,
    optionally scheduling projects the candidate already chose to build."""

    skills: list[str]
    target_role_id: str | None = None
    included_projects: list[SuggestedProject] = []


class LearningRoadmapCache(BaseModel):
    """Persisted roadmap (survives container restarts) - mirrors
    :class:`SuggestedProjectsCache`."""

    roadmap: LearningRoadmap | None = None
    cached: bool = False


class SuggestedProjectsRequest(BaseModel):
    """Regenerate portfolio projects from an explicit skill set."""

    skills: list[str]


class SuggestedProjectsCache(BaseModel):
    """Persisted project suggestions (survives container restarts)."""

    suggested_projects: list[SuggestedProject] = []
    skills: list[str] = []
    cached: bool = False


class SkillAnalysisResponse(BaseModel):
    """Fast, deterministic skill-gap result - see
    :class:`app.pipelines.skill_analysis_pipeline.SkillAnalysisPipeline`.
    Suggested projects and the learning roadmap are separate, explicit
    generations (see ``/suggested-projects`` and ``/learning-roadmap``),
    not part of this response.
    """

    core_strengths: list[str] = []
    skill_gaps: list[SkillGap] = []
    target_role_ready: bool = True
    target_role_title: str | None = None
    sample_status: str | None = None
    sample_job_count: int = 0
    error: str | None = None
